src/lsp/cobol_lsp/lsp_completion.ml

```diff
@@ -97,73 +54,69 @@ let range (pos:Position.t) text =
+      then (line_start, line_end)
+      else inner (line_end+1) (i+1)
+    in
+    let (start,end_) = inner 0 0 in
+    String.sub s start (end_-start)
+  in
+  let text_line = get_nthline (Lsp.Text_document.text text) line in
+  let index = List.fold_left (fun acc delimiter ->
       let current =
         try 1 + String.rindex_from text_line (character - 1) delimiter
         with _ -> 0 in
       max acc current) 0 delimiters in
-    let position_start = Position.create ~character:index ~line in
-    Range.create ~start:position_start ~end_:pos
+  let position_start = Position.create ~character:index ~line in
+  Range.create ~start:position_start ~end_:pos
 
 let to_completion_item ~kind ~range qualnames =
   List.flatten @@ List.map (function
-    | Cobol_ptree.Name name -> [~&name]
-    | Qual (name,_) as qualname ->
+      | Cobol_ptree.Name name -> [~&name]
+      | Qual (name,_) as qualname ->
         [~&name; Pretty.to_string "%a" Cobol_ptree.pp_qualname qualname]
-    | _ -> []) qualnames |>
-    List.map @@ completion_item ~kind ~range
+      | _ -> []) qualnames |>
+  List.map @@ completion_item ~kind ~range
+
+let string_of_K tokens =
+  let pp ppf =
+    let rec inner ?(space_before=true) = function
+      | [] -> ()
+      | hd::tl ->
+        let token' = hd &@ Srcloc.dummy in
+        if hd <> Cobol_parser.Tokens.PERIOD && space_before then Fmt.sp ppf ();
+        Cobol_parser.INTERNAL.pp_token ppf token';
+        inner tl
+    in inner ~space_before:false
+  in Pretty.to_string "%a" pp (Basics.NEL.to_list tokens)
 
 let map_completion_items ~(range:Range.t) ~group ~filename comp_entries =
-      let pos = range.end_ in
-      List.flatten @@ List.map (function
-        | Expect.QualifiedRef ->
-            to_completion_item
-              ~kind:CompletionItemKind.Variable ~range
-              (qualname_proposals ~filename pos group)
-        | ProcedureRef ->
-            to_completion_item
-            ~kind:CompletionItemKind.Module ~range
-            (procedure_proposals ~filename range.end_ group)
-        | K tokens -> begin
-            let tokens' = List.map (fun t -> t &@ Srcloc.dummy) @@ Basics.NEL.to_list tokens in
-            try let tokens = Pretty.to_string "%a" (Fmt.list ~sep:Fmt.sp Cobol_parser.INTERNAL.pp_token) tokens' in
-              [completion_item ~kind:CompletionItemKind.Keyword ~range tokens]
-            with Not_found -> [] end)
-      (Expect.CompEntrySet.elements comp_entries)
-
-(** [listpop l i] pops [i] elements of the list [l]
-    @return (h, t) where h is the list of popped element, t is the tail of the list *)
-let listpop l i =
-  let rec inner h t i =
-    if i == 0 then (h, t) else
-    match t with
-    | [] -> (h, [])
-    | hd::tl -> inner (hd::h) tl (i-1)
-  in let h, t = inner [] l i in List.rev h, t
-
-let pp_state ppf state = (* for debug *)
-  Fmt.pf ppf "%d" (Menhir.number state)
+  let pos = range.end_ in
+  List.flatten @@ List.map (function
```

Comment:
```suggestion
  List.flatten @@ List.rev_map (function
```
(if order is irrelevant at this point)

Not sure what `to_completion_item` does. It probably needs a better name.